dvb/datascience/pipeline.py

```diff
@@ -21,6 +24,63 @@ class Status(Enum):
     FINISHED = 3
 
 
+class TransformOutputsBase(metaclass=abc.ABCMeta):
+
+    @abc.abstractmethod
+    def save_pipe_output(self, pipe_name: str, transform_nr: int, output: Dict[str, Any]) ->None:
+        """save the pipe output"""
+
+    @abc.abstractmethod
+    def load_pipe_output(self, pipe_name: str, transform_nr: int) ->  Dict[str, Any]:
+        """load the pipe outputs, returns None when the output does not exists"""
+
+    @abc.abstractmethod
+    def contains_pipe(self, pipe_name: str, transform_nr: int) -> bool:
+        """returns True when the pipe with pipe_name has stored his output.
+        For some implementations this is more efficient as checking whether load_pipe_output is not None
+        """
+
+class TransformOutputsMemory(TransformOutputsBase):
+
+    # outputs = None
+
+    def __init__(self):
+        self.outputs: Dict[Tuple[str, int], Dict[str, Any]] = {}
+
+    def save_pipe_output(self, pipe_name: str, transform_nr: int, output: Dict[str, Any]) ->None:
+        self.outputs[(pipe_name, transform_nr)] = output
+
+    def load_pipe_output(self, pipe_name: str, transform_nr: int) -> Dict[str, Any]:
+        return self.outputs[(pipe_name, transform_nr)]
+
+    def contains_pipe(self, pipe_name: str, transform_nr: int) -> bool:
+        return (pipe_name, transform_nr) in self.outputs
+
+class TransformOutputsFile(TransformOutputsBase):
+
+    def __init__(self, directory='transform_output', file_pattern='{pipe_name}_{transform_nr}.pkl'):
+        self.directory = pathlib.Path(directory)
+        self.file_pattern = file_pattern
+
+        self.directory.mkdir(parents=True, exist_ok=True)
+
+    def save_pipe_output(self, pipe_name: str, transform_nr: int,  output: Dict[str, Any]) ->None:
+        filename = self.file_pattern.format(pipe_name=pipe_name, transform_nr= transform_nr)
+        with open(self.directory / filename, 'wb') as f:
+            pickle.dump(output, f)
+
+    def load_pipe_output(self, pipe_name: str, transform_nr: int) -> Dict[str, Any]:
+        filename = self.file_pattern.format(pipe_name=pipe_name, transform_nr= transform_nr)
```

Comment:
```suggestion
        filename = self.file_pattern.format(pipe_name=pipe_name, transform_nr=transform_nr)
```
(removed space before transform_nr)